sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/package.scala

```diff
@@ -169,25 +181,50 @@ package object expressions  {
         })
       }
 
-      // Find matches for the given name assuming that the 1st part is a qualifier (i.e. table name,
-      // alias, or subquery alias) and the 2nd part is the actual name. This returns a tuple of
+      // Find matches for the given name assuming that the 1st two parts are qualifier
+      // (i.e. database name and table name) and the 3rd part is the actual column name.
+      //
+      // For example, consider an example where "db1" is the database name, "a" is the table name
+      // and "b" is the column name and "c" is the struct field name.
+      // If the name parts is db1.a.b.c, then Attribute will match
```

Comment:
we may need to check the SQL standard about how to resolve the ambiguity. For `a.b.c`, it's possible that `a` is db name, `b` is table name, `c` is column name. If there is no column named `c`, we should fail. It's also possible that `a` is table name, `b` is column name, `c` is nested field name, and it exists. What's the expected behavior here?

Reply:
In this case, if a.b.c fails to resolve as db.table.column, then we check if there is a table  and column that matches a.b and then see if c is a nested field name and if it exists, it will resolve to the nested field. 

Tests with struct nested fields are [here](https://github.com/skambha/spark/blob/90cd6d33f59fdae16a3a386ed14cefb3f28d35a8/sql/core/src/test/resources/sql-tests/inputs/columnresolution.sql#L65-L73)

Reply:
Is this behavior defined by SQL standard? IIRC the current behavior without your patch is
`col.innerField1.innerField2` can fail even we have a table named `col` and a column named `innerField1` which has a `innerField2` field.

Reply:
Actually it won't fail without this patch too,  the col.innerField1.innerField2 will resolve correctly to the struct field if there is a table named col and column named innerField1 which has a innerField2 field.   

Please see an example of such a scenario in this test output on master spark: 
https://github.com/apache/spark/blob/master/sql/core/src/test/resources/sql-tests/results/columnresolution.sql.out#L360
------
Let me look into the sql standard and get back on that.

Reply:
What I'm talking about is ambiguity. `col.innerField1.innerField2` can fail if the `innerField2` doesn't exist. My question is: shall we try all the possible resolution paths and pick the valid one, or define a rule that we can decide the resolution path ahead.

Reply:
@cloud-fan , Thank you for your suggestion and question. 

Existing spark behavior follows precedence rules in the column resolution logic and in this patch we are following the same pattern/rule.   

I am looking into the SQL standard to see if there are any column resolution rules but I have not found any yet.  However when I researched existing databases, I observed different behaviors among them and it is listed in Section 2/Table A in the design doc [here](https://drive.google.com/file/d/1zKm3aNZ3DpsqIuoMvRsf0kkDkXsAasxH/view).

I agree, we can improve upon the checks in existing precedence to go all the way to ensure there is a nested field.  Although, the user can always qualify the field to resolve the ambiguity.  Shall we open another issue to discuss and improve upon the existing resolution logic.